src/AutoMapper/Configuration/Conventions/SourceToDestinationNameMapperAttributesMember.cs

```diff
@@ -8,15 +7,18 @@ namespace AutoMapper.Configuration.Conventions
     public class SourceToDestinationNameMapperAttributesMember : ISourceToDestinationNameMapper
     {
         private static readonly SourceMember[] Empty = new SourceMember[0];
-        private readonly Dictionary<TypeDetails, SourceMember[]> _allSourceMembers = new Dictionary<TypeDetails, SourceMember[]>();
+        private LockingConcurrentDictionary<TypeDetails, SourceMember[]> _allSourceMembers
+            = new LockingConcurrentDictionary<TypeDetails, SourceMember[]>(_ => Empty);
```

Comment:
I don't think we should lock here. It's difficult to write correct thread safe code in a lot of places. I think we should lock when calling CreateMap.

Reply:
We already lock there but this is another place that's a static cache. Another option would be just to get rid of the cache.

Reply:
The idea was to not look for the attribute every time it tries to match the members to map. So it will be slower without the cache. But my point point is more general. A coarser grained lock would make the code easier to write and understand. That's why I added a global lock before each CreateMap. So we don't have to worry about concurrency at every point.